### PR TITLE
PersonalAttributions: add config and add to API to complete feature

### PR DESCRIPTION
**Pull request by blueridger**
# Description
This adds a config to the instance for the Personal Attribution feature. It also uses the config in the credrank API. This completes the Personal Attribution feature V1.

# Test Plan
Edited the `config/personalAttributions.json` and then ran `scdev credrank -d` and verified console output as well as json reformatting.
## No config file
Console output:
```

# Top Participants By New Cred
┌─────────┬──────────────────┬────────────┬───────────┬──────────┐
│ (index) │       Name       │ Prior Cred │ New Cred  │ % Change │
├─────────┼──────────────────┼────────────┼───────────┼──────────┤
│    0    │  'decentralion'  │ '37739.3'  │ '37739.3' │  '0.0%'  │
│    1    │   'lbstrobbe'    │ '17159.3'  │ '17159.3' │  '0.0%'  │
│    2    │     'Thena'      │ '16911.7'  │ '16911.7' │  '0.0%'  │
│    3    │     's-ben'      │ '14595.6'  │ '14595.6' │  '0.0%'  │
│    4    │    'KuraFire'    │ '14573.2'  │ '14573.2' │  '0.0%'  │
│    5    │    'wchargin'    │ '13000.1'  │ '13000.1' │  '0.0%'  │
│    6    │   'topocount'    │ '12720.9'  │ '12720.9' │  '0.0%'  │
│    7    │  'panchomiguel'  │ '12141.9'  │ '12141.9' │  '0.0%'  │
│    8    │     'hammad'     │ '11310.1'  │ '11310.1' │  '0.0%'  │
│    9    │  'joiecousins'   │  '9122.3'  │ '9122.3'  │  '0.0%'  │
│   10    │      'Jojo'      │  '8023.5'  │ '8023.5'  │  '0.0%'  │
│   11    │      'bex'       │  '7134.5'  │ '7134.5'  │  '0.0%'  │
│   12    │   'sandpiper'    │  '6903.1'  │ '6903.1'  │  '0.0%'  │
│   13    │     'beanow'     │  '6524.2'  │ '6524.2'  │  '0.0%'  │
│   14    │   'echojuliet'   │  '5063.7'  │ '5063.7'  │  '0.0%'  │
│   15    │   'dependabot'   │  '2811.3'  │ '2811.3'  │  '0.0%'  │
│   16    │    'benoxmo'     │  '2515.5'  │ '2515.5'  │  '0.0%'  │
│   17    │  'saintmedusa'   │  '2372.9'  │ '2372.9'  │  '0.0%'  │
│   18    │    'burrrata'    │  '2006.2'  │ '2006.2'  │  '0.0%'  │
│   19    │ 'BrianBelhumeur' │  '1997.9'  │ '1997.9'  │  '0.0%'  │
└─────────┴──────────────────┴────────────┴───────────┴──────────┘
```
Config file created as output:
```
[
]
```
## With test file
Test file:
```
[
  {
    "fromParticipantName": "Thena",
    "recipients": [
      {
        "toParticipantName": "Jojo",
        "proportions": [
          {
            "startDate": "2/15/21",
            "decimalProportion": 0.1
          }
        ]
      }
    ]
  }
]
```
Console output:
note: other people went up a little and jojo went up not so much because Jojo sends more emojis than me, lol yikes
```
# Top Participants By New Cred
┌─────────┬──────────────────┬────────────┬───────────┬──────────┐
│ (index) │       Name       │ Prior Cred │ New Cred  │ % Change │
├─────────┼──────────────────┼────────────┼───────────┼──────────┤
│    0    │  'decentralion'  │ '37739.3'  │ '37775.2' │  '0.1%'  │
│    1    │   'lbstrobbe'    │ '17159.3'  │ '17178.8' │  '0.1%'  │
│    2    │     'Thena'      │ '16911.7'  │ '16219.3' │ '−4.1%'  │
│    3    │     's-ben'      │ '14595.6'  │ '14616.4' │  '0.1%'  │
│    4    │    'KuraFire'    │ '14573.2'  │ '14594.0' │  '0.1%'  │
│    5    │    'wchargin'    │ '13000.1'  │ '13012.0' │  '0.1%'  │
│    6    │   'topocount'    │ '12720.9'  │ '12740.7' │  '0.2%'  │
│    7    │  'panchomiguel'  │ '12141.9'  │ '12161.4' │  '0.2%'  │
│    8    │     'hammad'     │ '11310.1'  │ '11321.0' │  '0.1%'  │
│    9    │  'joiecousins'   │  '9122.3'  │ '9134.4'  │  '0.1%'  │
│   10    │      'Jojo'      │  '8023.5'  │ '8409.2'  │  '4.8%'  │
│   11    │      'bex'       │  '7134.5'  │ '7142.2'  │  '0.1%'  │
│   12    │   'sandpiper'    │  '6903.1'  │ '6911.3'  │  '0.1%'  │
│   13    │     'beanow'     │  '6524.2'  │ '6529.5'  │  '0.1%'  │
│   14    │   'echojuliet'   │  '5063.7'  │ '5079.8'  │  '0.3%'  │
│   15    │   'dependabot'   │  '2811.3'  │ '2813.7'  │  '0.1%'  │
│   16    │    'benoxmo'     │  '2515.5'  │ '2517.8'  │  '0.1%'  │
│   17    │  'saintmedusa'   │  '2372.9'  │ '2379.2'  │  '0.3%'  │
│   18    │    'burrrata'    │  '2006.2'  │ '2007.8'  │  '0.1%'  │
│   19    │ 'BrianBelhumeur' │  '1997.9'  │ '2002.4'  │  '0.2%'  │
└─────────┴──────────────────┴────────────┴───────────┴──────────┘
```
Config file overwritten as output:
```
[
    {
        "fromParticipantId": "6wyPOVWLmZ2b35eCmOrz1w",
        "fromParticipantName": "Thena",
        "recipients": [
            {
                "proportions": [
                    {
                        "decimalProportion": 0.1,
                        "startDate": "2/15/21"
                    }
                ],
                "toParticipantId": "8vvdXoegkQ7LLAfdrQ333Q",
                "toParticipantName": "Jojo"
            }
        ]
    }
]
```
## After merging my identity into another
Config file overwritten as ouput:
```
[
    {
        "fromParticipantId": "u4A1GCZPtGPnyD0q1dsRTQ",
        "fromParticipantName": "thena-test",
        "recipients": [
            {
                "proportions": [
                    {
                        "decimalProportion": 0.1,
                        "startDate": "2/15/21"
                    }
                ],
                "toParticipantId": "8vvdXoegkQ7LLAfdrQ333Q",
                "toParticipantName": "Jojo"
            }
        ]
    }
]
```

### CODE REVIEW

File: src/api/config/personalAttributionsConfig.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+      +startDate: TimestampISO,
+      // The 0-1 representation of the percentage of cred that should flow from
+      // the fromParticipant to the toParticipant each epoch
+      +decimalProportion: number,
+    |}>,
+  |}>,
+|}>;
+
+/**
+  Adds the IdentityIds where only IdentityNames are provide, and updates names
```
Comment (topocount) [inline src/api/config/personalAttributionsConfig.js:53]:
```suggestion
  Adds the IdentityIds where only IdentityNames are provided, and updates names
```

File: src/api/config/personalAttributionsConfig.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+} from "../../util/timestamp";
+import * as C from "../../util/combo";
+import {type PersonalAttributions} from "../../core/credrank/personalAttribution";
+
+/**
+  This type resembles the JSON schema for configuring personal attributions,
+  which allows participants to attribute their cred to other participants.
+  This feature should not be used to make cred sellable/transferable, but
+  instead is intended to allow participants to acknowledge that a portion of
+  their creditted outputs are directly generated/supported by the labor of
```
Comment (topocount) [inline src/api/config/personalAttributionsConfig.js:23]:
```suggestion
  their credited outputs are directly generated/supported by the labor of
```

File: src/api/main/credrank.js [hunk 3]
```diff
@@ -44,13 +51,6 @@ export async function credrank(input: CredrankInput): Promise<CredrankOutput> {
     merge(input.pluginGraphs),
     input.weightOverrides
   );
-  const dependenciesWithIds = input.dependencies.map((d) =>
-    // This mutates the ledger, adding new identities when needed.
-    ensureIdentityExists(d, input.ledger)
-  );
-  const dependencyBonuses = dependenciesWithIds.map((d) =>
-    toBonusPolicy(d, input.ledger)
-  );
```
Comment (topocount) [inline src/api/main/credrank.js:53]: why move this in this commit? what does this have to do with the attribution config?

File: src/api/config/personalAttributionsConfig.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  others. (e.g. when a contributor has a personal assistant working behind the 
+  scenes)
+ */
+export type PersonalAttributionsConfig = Array<{|
+  // The name of a participant that is flowing cred to others
+  fromParticipantName: Name,
+  // The id of the participant. Will be autopopulated if absent.
+  fromParticipantId?: IdentityId,
```
Comment (topocount) [inline src/api/config/personalAttributionsConfig.js:31]: I understand that keying on the participant name is tenuous, but I think exposing both in a config should be handled more rigorously. If both are present, both should be validated, or the config should error as invalid if an account merge doesn't point towards a clear update for both values. In the case that the `IdentityId` was merged into another Identity, the name needs to be updated as well, since it's a mandatory config.

if we are going to update config values automatically, we should avoid letting the name go stale, basically.

File: src/api/config/personalAttributionsConfig.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  +recipients: Array<{|
+    // The name of a participant that is receiving cred
+    // from the "fromParticipant"
+    toParticipantName: Name,
+    // The id of the participant. Will be autopopulated if absent.
+    toParticipantId?: IdentityId,
```
Comment (topocount) [inline src/api/config/personalAttributionsConfig.js:37]: Same comment as above. It might make sense to write a `participantValidator` or `participantUpdater` function that updates this and the `from` values according to the logic above.

File: src/api/config/personalAttributionsConfig.js (hunk 4 above)
Comment (blueridger) [reply src/api/config/personalAttributionsConfig.js:31]: The name is updated automatically already. See line 65. And because it can go stale and merges/renames don't store records of previously names, keying on name after the ID has been added to the config isn't feasible. I would say we could remove the name after replacing it with ID, but then edits are harder for human users.

File: src/api/config/personalAttributionsConfig.js (hunk 5 above)
Comment (blueridger) [reply src/api/config/personalAttributionsConfig.js:37]: Literally what updatePersonalAttributionsConfig does.

File: src/api/main/credrank.js (hunk 3 above)
Comment (blueridger) [reply src/api/main/credrank.js:53]: Just baking in a little organization of the method by putting all the dependency-related code together.

File: src/api/config/personalAttributionsConfig.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+      +decimalProportion: number,
+    |}>,
+  |}>,
+|}>;
+
+/**
+  Adds the IdentityIds where only IdentityNames are provide, and updates names
+  and ids to reflect the account's current identity after merging/renaming.
+ */
+export function updatePersonalAttributionsConfig(
```
[the remark this answers is not in the document]
Comment (topocount) [reply src/api/config/personalAttributionsConfig.js:56]: I mean to delete this one :grimacing:

**Review by topocount — APPROVED** [review]
